src/System.CommandLine.StaticCompletions/shells/PowershellShellProvider.cs

```diff
@@ -0,0 +1,248 @@
+        writer.Indent++;
+        GenerateSubcommandCompletions([], writer, command);
+
+        writer.Indent--;
+        writer.WriteLine("}");
+        writer.WriteLine("$completions | Where-Object -FilterScript { $_.CompletionText -like \"$wordToComplete*\" } | Sort-Object -Property ListItemText");
+
+        writer.Indent--;
+        writer.WriteLine("}");
+        writer.Flush();
+        return textWriter.ToString();
+    }
+
+    private static string CompletionResult(string name, string value, string type, string? helpText)
+    {
+        if (helpText is null)
+        {
+            return $"[CompletionResult]::new('{value}', '{name}', [CompletionResultType]::{type}, \"{value}\")";
+        }
+        else
+        {
+            return $"[CompletionResult]::new('{value}', '{name}', [CompletionResultType]::{type}, \"{helpText}\")";
+        }
+    }
+
+    private static string ParameterNameResult(string name, string value, string? helpText) => CompletionResult(name, value, "ParameterName", helpText);
+
+    private static string ParameterValueResult(string name, string value, string? helpText) => CompletionResult(name, value, "ParameterValue", helpText);
+
+    private static string? SanitizeHelpDescription(CliSymbol s) => s.Description?.ReplaceLineEndings(" ").Replace("`", "``").Replace("'", "`'").Replace("\"", "`\"").Replace("$", "`$");
+
+    /// <summary>
+    /// Generations completion-list items for the names of the given option. Typically used by commands/subcommands for static lookup lists.
+    /// </summary>
+    /// <param name="o"></param>
+    /// <returns></returns>
+    private static IEnumerable<string> GenerateOptionNameCompletions(CliOption o)
+    {
+        if (o.Hidden)
+        {
+            yield break;
+        }
+
+        var verboseName = o.Name;
+        var names = o.Names();
+        var helpText = SanitizeHelpDescription(o);
+
+        // generate a completion recognizer for each alias, but have it's 'commit' value
+        // be the longest/most verbose for clarity
+        foreach (var name in names)
+        {
+            // differentiate casing for single character flags because powershell doesn't do case-sensitivity
+            var completionValue = name.IsUpperCaseSingleCharacterFlag() ? $"{verboseName} " : verboseName;
+            yield return ParameterNameResult(name, completionValue, helpText);
+        }
+    }
+
+    /// <summary>
+    /// Generate completions for the statically-known arguments
+    /// </summary>
+    /// <param name="argument"></param>
+    /// <returns></returns>
+    private static IEnumerable<string> GenerateArgumentCompletions(CliArgument argument)
+    {
+        if (argument.Hidden)
+        {
+            yield break;
+        }
+
+        if (argument.IsDynamic())
+        {
+            // if the argument is a not-static-friendly argument, we need to call into the app for completions
+            // TODO: not yet supported for powershell
```

Comment:
Hmm. I just looked and I thought this feature existed: https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_functions_argument_completion?view=powershell-7.5#dynamic-validateset-values-using-classes

Reply:
Nice call out - we could maybe version-probe here? I'd be willing to leave that as an enhancement though

Reply:
Yeah, I'd just make a new issue to add this feature later.